### PR TITLE
1117 jump to main content

### LINKED ISSUE #1117: Accesibility: missing "jump to content" link
### Issue Summary

There's no "skip to content" link allowing users with screenreaders to skip over all the headers.

### Steps to Reproduce

Go to the main page of any lemmy site.  Click tab.  It should take you to a "jump to content" link, but none appears.

The screenshot of Reddit's interface illustrates what should happen. The "jump to content" link in the image is initally invisible, but appears when I hit a tab.

Screen Shot 2023-06-08 at 11 37 17 AM

### Technical details

https://www.useragentman.com/blog/2016/12/15/implementing-an-accessible-skip-navigation-link-requires-more-thought-than-youd-think/  is a good discussion of how to provide this functionality (although a little old, not sure if best practices have evolved since then)

### PR DESCRIPTION
**Pull request by skspade**
# Describe 
- Adds ability to press the "Tab" key to reveal a hidden "Jump to conent" link. Clicking or hitting enter will bring the post listings regardless of page into focus. 
- Prevents `validURL` from throwing exception into the browser console.

# Video 

https://github.com/LemmyNet/lemmy-ui/assets/43147936/203ce936-8070-4ea3-8100-0be18233f3ad

# Issue 
closes #1117

### CODE REVIEW

File: src/assets/css/main.css [hunk 1]
```diff
@@ -430,3 +430,19 @@ br.big {
 em-emoji-picker {
   width: 100%;
 }
+
+.skip-link {
+  position: absolute;
+  top: -40px;
+  left: 0;
+  background: #000;
```
Comment (SleeplessOne1917) [inline src/assets/css/main.css:438]: Could you please use bootstrap class names for styling where possible? You could give the skip link the following classes: "bg-dark text-light p-2 text-decoration-none position-absolute start-0 z-3" The only properties you'll still need defined on this css class are top and transition.

File: src/server/index.tsx [hunk 2]
```diff
@@ -393,6 +393,13 @@ async function createSsrHtml(root: string, isoData: IsoDataOptionalSite) {
   </head>
 
   <body ${helmet.bodyAttributes.toString()}>
+    <a
+    href="#main-content"
+    class="skip-link"
+    onclick="document.getElementById('main-content').focus(); return false;"
```
Comment (SleeplessOne1917) [inline src/server/index.tsx:399]: You don't have to make this function a string. You could make it:

```tsx
<a onClick={() => {
    document.getElementById('main-content')?.focus(); 
}}
>
    Jump to content
</a>
```

File: src/server/index.tsx [hunk 3]
```diff
@@ -393,6 +393,13 @@ async function createSsrHtml(root: string, isoData: IsoDataOptionalSite) {
...
+    >
+      Jump to content
```
Comment (SleeplessOne1917) [inline src/server/index.tsx:401]: This is going to need a translation. [Please add one to the translations repo](https://github.com/LemmyNet/lemmy-translations).

**Review by SleeplessOne1917 — CHANGES_REQUESTED** [review]
Please address the changes I requested in the other comments. In addition to that, this jump to content link should be able to work in more places than just the post listings. As it is now, this will only allow users to jump to content on their home screen or a community screen.

**Comment by skspade** [comment]
Great! I'll get to work on your suggestions tonight. Thanks for the review

File: src/server/index.tsx (hunk 2 above)
Comment (skspade) [reply src/server/index.tsx:399]: Are you able to use JSX syntax here? This is in the `createSsrHtml` function which does not seem to be JSX but a HTML in a template string.

Comment (SleeplessOne1917) [reply src/server/index.tsx:399]: That is a good point. I didn't realize this wasn't JSX in the small window github gave me.

**Comment by skspade** [comment]
> This is going to need a translation. [Please add one to the translations repo](https://github.com/LemmyNet/lemmy-translations). 

https://github.com/LemmyNet/lemmy-translations/pull/50

File: src/assets/css/main.css (hunk 1 above)
Comment (skspade) [reply src/assets/css/main.css]: The bootstrap color classes are confusing me a bit. For example `bg-dark` is giving me a light color background. `text-light` is giving me dark text. The opposite is true too. Look at my screenshots below.
Screenshot from 2023-06-11 13-11-04
Screenshot from 2023-06-11 13-11-29

**Comment by skspade** [comment]
I re-implemented the feature, this time I wrapped all routes at the entry point. I feel like this is effectively skips the navigation elements on every page and in the future other developers do not have to worry about wrapping new pages with an element containing the attribute `main-content`

**Review by dessalines — COMMENTED** [review]
Externalize that one to a handler function. The convention is to have any onClick actions use linkEvent with a function.

**Comment by skspade** [comment]
Sounds good, I can move it into a JSX portion of the code. I put it here since it needs to be the first focusable element on any page. But I agree consistency is important. I'll get to work on this in the evening

**Comment by skspade** [comment]
Updated Video showing the functionality is the same when written in JSX
https://github.com/LemmyNet/lemmy-ui/assets/43147936/62b5f0eb-6a22-49f6-bd3c-30e718bf21f4

**Comment by skspade** [comment]
https://github.com/LemmyNet/lemmy-ui/assets/43147936/f0991bfa-63d5-4ed7-9f1a-c648e525f2f2

**Comment by skspade** [comment]
I see that there is conflicts now, I'll fix those tonight

**Comment by SleeplessOne1917** [comment]
@skspade The only conflict is the translations submodule. Try running `git submodule update --remote --recursive` and committing that.

File: src/shared/utils.ts [hunk 4]
```diff
@@ -329,7 +329,12 @@ export function isVideo(url: string) {
 }
 
 export function validURL(str: string) {
-  return !!new URL(str);
+  try {
+    new URL(str);
+    return true;
+  } catch (_) {
```
Comment (ayan4m1) [inline src/shared/utils.ts:335]: The discard parameter can be omitted entirely here

```jsx
try {
  new URL(str);
  return true;
} catch {
  return false;
}
```

Comment (skspade) [reply src/shared/utils.ts:335]: Good point! I'll make that change too

File: src/assets/css/main.css [hunk 5]
```diff
@@ -434,3 +434,12 @@ br.big {
 em-emoji-picker {
   width: 100%;
 }
+
+.skip-link {
+  top: -40px;
+  transition: top 0.3s ease;
```
[the remark this answers is not in the document]
Comment (skspade) [reply src/assets/css/main.css:440]: True! Since we do not have a user preference for this, I think Its best to just remove the transition.

**Comment by skspade** [comment]
@SleeplessOne1917  I believe I got it to a good place. Let me know if you would like me to squash my commits before we merge (if you approve)

File: src/assets/css/main.css (hunk 5 above)
[the remark this answers is not in the document]
Comment (ayan4m1) [reply src/assets/css/main.css]: Oh, I meant using a media query: https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion

[the remark this answers is not in the document]
Comment (skspade) [reply src/assets/css/main.css]: Hey thanks for that! Never knew you could query for that! 
I took up this PR to learn more about accessibility so thanks for spreading some knowledge

**Comment by SleeplessOne1917** [comment]
@skspade  Squashing is fine.

**Review by SleeplessOne1917 — APPROVED** [review]

**Comment by skspade** [comment]
@SleeplessOne1917 Thanks for the approval! Looks like woodpecker is stalling out after I squashed.  Rebased commit: ` 36033b7` passed, then I squashed to ` d3e1812 `  which doesn't seem to be detected.

**Pull request merged**